dipy/reconst/dki_micro.py

```diff
@@ -0,0 +1,507 @@
+    B = np.zeros((sphere.x.size, 6))
+    B[:, 0] = sphere.x * sphere.x  # Bxx
+    B[:, 1] = sphere.x * sphere.y * 2.  # Bxy
+    B[:, 2] = sphere.y * sphere.y   # Byy
+    B[:, 3] = sphere.x * sphere.z * 2.  # Bxz
+    B[:, 4] = sphere.y * sphere.z * 2.  # Byz
+    B[:, 5] = sphere.z * sphere.z  # Bzz
+    pinvB = np.linalg.pinv(B)
+
+    # Compute hindered and restricted diffusion tensors for all voxels
+    evals, evecs, kt = split_dki_param(dki_params)
+    dt = lower_triangular(vec_val_vect(evecs, evals))
+    md = mean_diffusivity(evals)
+
+    index = ndindex(mask.shape)
+    for idx in index:
+        if not mask[idx]:
+            continue
+        # sample apparent diffusion and kurtosis values
+        di = directional_diffusion(dt[idx], sphere.vertices)
+        ki = directional_kurtosis(dt[idx], md[idx], kt[idx], sphere.vertices,
+                                  adc=di)
+        edi = di * (1 + np.sqrt(ki * awf[idx] / (3.0 - 3.0 * awf[idx])))
+        edt = np.dot(pinvB, edi)
+        edt_all[idx] = edt
+
+        # We only move on if there is an axonal water fraction.
+        # Otherwise, remaining params are already zero, so move on
+        if awf[idx] == 0:
+            continue
+        # Convert apparent diffusion and kurtosis values to apparent diffusion
+        # values of the hindered and restricted diffusion
+        idi = di * (1 - np.sqrt(ki * (1.0 - awf[idx]) / (3.0 * awf[idx])))
+        # generate hindered and restricted diffusion tensors
+        idt = np.dot(pinvB, idi)
+        idt_all[idx] = idt
+
+    return edt_all, idt_all
+
+
+def dkimicro_prediction(params, gtab, S0=1):
+    r""" Signal prediction given the DKI microstructure model parameters.
+
+    Parameters
+    ----------
+    params : ndarray (x, y, z, 40) or (n, 40)
+    All parameters estimated from the diffusion kurtosis microstructure model.
+        Parameters are ordered as follows:
+            1) Three diffusion tensor's eigenvalues
+            2) Three lines of the eigenvector matrix each containing the
+               first, second and third coordinates of the eigenvector
+            3) Fifteen elements of the kurtosis tensor
+            4) Six elements of the hindered diffusion tensor
+            5) Six elements of the restricted diffusion tensor
+            6) Axonal water fraction
+    gtab : a GradientTable class instance
+        The gradient table for this prediction
+    S0 : float or ndarray
+        The non diffusion-weighted signal in every voxel, or across all
+        voxels. Default: 1
+
+    Returns
+    --------
+    S : (..., N) ndarray
+        Simulated signal based on the DKI microstructure model
+
+    Notes
+    -----
+    The predicted signal is given by:
+    $S(\theta, b) = S_0 * [f * e^{-b ADC_{r}} + (1-f) * e^{-b ADC_{h}]$, where
+    $ ADC_{r} and ADC_{h} are the apparent diffusion coefficients of the
+    diffusion hindered and restricted compartment for a given direction
```

Comment:
Maybe this is the place to mention that these are referred to as "intracellular" and "extracellular" in the Fieremans paper.